File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/shapes/custom_question_controller.py
import logging
import inspect
from django.contrib import messages
from django.http import HttpRequest, HttpResponse
from django.shortcuts import render, redirect, get_object_or_404
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._traits.controller import Controller
from ....Models.shapes.custom_question import CustomQuestion

logger = logging.getLogger(__name__)

class CustomQuestionController(Controller):

    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('manage custom question'):
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            qs = CustomQuestion.objects.filter(created_by=request.user.creator_id())
            return render(request, 'customQuestion/index.html', {'questions': qs})
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            return render(request, 'customQuestion/create.html', {'is_required': CustomQuestion.is_required})
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def store(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('create custom question'):
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            data = request.POST
            if not data.get('question'):
                messages.error(request, 'Validation Error: question required.')
                return redirect(get_redirect_url(request))
            cq = CustomQuestion()
            cls._set_custom_question(cq, data, request.user.creator_id())
            messages.success(request, 'Question successfully created.')
            return redirect(get_redirect_url(request))
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def show(cls, request: HttpRequest, custom_question_id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            return redirect('custom_question_index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def edit(cls, request: HttpRequest, custom_question_id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            cq = get_object_or_404(CustomQuestion, pk=custom_question_id)
            return render(request, 'customQuestion/edit.html', {
                'custom_question': cq,
                'is_required': CustomQuestion.is_required
            })
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def update(cls, request: HttpRequest, custom_question_id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('edit custom question'):
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            data = request.POST
            if not data.get('question'):
                messages.error(request, 'Validation Error: question required.')
                return redirect(get_redirect_url(request))
            cq = get_object_or_404(CustomQuestion, pk=custom_question_id)
            cls._set_custom_question(cq, data, None)
            messages.success(request, 'Question successfully updated.')
            return redirect(get_redirect_url(request))
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def destroy(cls, request: HttpRequest, custom_question_id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('delete custom question'):
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            cq = get_object_or_404(CustomQuestion, pk=custom_question_id)
            cq.delete()
            messages.success(request, 'Question successfully deleted.')
            return redirect(get_redirect_url(request))
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def _set_custom_question(cls, obj: CustomQuestion, data, user_id=None):
        obj.question = data.get('question')
        obj.is_required = data.get('is_required')
        if user_id is not None:
            obj.created_by = user_id
        obj.save()

File: _inc/laravel/app/Http/Controllers/Shapes/CustomQuestionController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    DatabaseConstants,
    MiddlewaresConstants,
    PermissionsConstants,
    ViewsConstants
};
use App\Models\CustomQuestion;
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{Request, RedirectResponse, JsonResponse};
use Illuminate\Support\Facades\{Auth, DB, Log, Validator, View as ViewFacade};
use Illuminate\View\View;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
class CustomQuestionController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    private const REDIRECT_ROUTE = ViewsConstants::CST_QT . '.index';

    public function __construct()
    {
        $this->middleware([MiddlewaresConstants::AUTH, MiddlewaresConstants::XSS]);
    }

    public function index(Request $request): View|RedirectResponse
    {
        $action = __FUNCTION__;
        $cls = static::class;
        $sig = "$cls::$action";
        return $this->measureProfile(function () use ($request, $action, $sig) {
            Log::info("$sig start", ['user' => Auth::id()]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, PermissionsConstants::MNG_CST_QT, self::REDIRECT_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            try {
                $t = microtime(true);
                $questions = CustomQuestion::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())->get();
                $this->logExecutionTime($t, "$sig::fetchQuestions", 'completed');

                $viewPath = ViewsConstants::CST_QT . '.index';
                $t = microtime(true);
                if (!ViewFacade::exists($viewPath)) {
                    $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');
                    return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                }
                $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');

                Log::info("$sig fetched", ['count' => $questions->count()]);
                return view($viewPath, compact('questions'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $sig);
            }
        });
    }

    public function create(Request $request): View|JsonResponse|RedirectResponse
    {
        $action = __FUNCTION__;
        $cls = static::class;
        $sig = "$cls::$action";
        return $this->measureProfile(function () use ($request, $action, $sig) {
            Log::info("$sig start", ['user' => Auth::id()]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($c = self::guard($request, 'create custom question', self::REDIRECT_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            try {
                $t = microtime(true);
                $isRequired = CustomQuestion::$is_required;
                $this->logExecutionTime($t, "$sig::prepareData", 'completed');

                $viewPath = ViewsConstants::CST_QT . '.create';
                $t = microtime(true);
                if (!ViewFacade::exists($viewPath)) {
                    $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');
                    return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                }
                $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');

                return view($viewPath, compact('isRequired'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $sig);
            }
        });
    }

    public function store(Request $request): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $cls = static::class;
        $sig = "$cls::$action";
        return $this->measureProfile(function () use ($request, $action, $sig) {
            Log::info("$sig start", ['input' => $request->only('question', 'is_required')]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'create custom question', self::REDIRECT_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }

            $t = microtime(true);
            $validator = Validator::make($request->all(), ['question' => 'required']);
            $this->logExecutionTime($t, "$sig::validate", 'completed');
            if ($validator->fails()) {
                $msg = $validator->errors()->first();
                Log::warning("$sig validation failed", ['error' => $msg]);
                return redirect()->back()->with('error', $msg);
            }

            try {
                $t = microtime(true);
                DB::transaction(function () use ($request, $user) {
                    $q = new CustomQuestion();
                    foreach (['question', 'is_required'] as $field) {
                        $q->$field = $request->input($field);
                    }
                    $q[DatabaseConstants::COL_TABLE_CREATOR] = $user?->creatorId();
                    $q->save();
                    Log::info('store: created', ['id' => $q->id]);
                });
                $this->logExecutionTime($t, "$sig::transaction", 'completed');

                return redirect()->route(self::REDIRECT_ROUTE)->with('success', __('Question successfully created.'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $sig);
            }
        });
    }

    public function show(CustomQuestion $customQuestion, Request $request): View|RedirectResponse
    {
        $action = __FUNCTION__;
        $cls = static::class;
        $sig = "$cls::$action";
        return $this->measureProfile(function () use ($customQuestion, $request, $action, $sig) {
            Log::info("$sig start", ['id' => $customQuestion->id]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($c = self::guard($request, 'view custom question', self::REDIRECT_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }

            try {
                $viewPath = ViewsConstants::CST_QT . '.show';
                $t = microtime(true);
                if (!ViewFacade::exists($viewPath)) {
                    $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');
                    return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                }
                $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');

                return view($viewPath, compact('customQuestion'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $sig);
            }
        });
    }

    public function edit(CustomQuestion $customQuestion, Request $request): View|RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $cls = static::class;
        $sig = "$cls::$action";
        return $this->measureProfile(function () use ($customQuestion, $request, $action, $sig) {
            Log::info("$sig start", ['id' => $customQuestion->id]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($c = self::guard($request, 'edit custom question', self::REDIRECT_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            if (!$this->isOwner($customQuestion)) {
                Log::warning("$sig not owner", ['user' => Auth::id()]);
                return defaultPermissionDenial($request, new AuthorizationException, $sig);
            }

            try {
                $t = microtime(true);
                $isRequired = CustomQuestion::$is_required;
                $this->logExecutionTime($t, "$sig::prepareData", 'completed');

                $viewPath = ViewsConstants::CST_QT . '.edit';
                $t = microtime(true);
                if (!ViewFacade::exists($viewPath)) {
                    $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');
                    return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                }
                $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');

                return view($viewPath, compact('customQuestion', 'isRequired'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $sig);
            }
        });
    }

    public function update(Request $request, CustomQuestion $customQuestion): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $cls = static::class;
        $sig = "$cls::$action";
        return $this->measureProfile(function () use ($request, $customQuestion, $action, $sig) {
            Log::info("$sig start", ['id' => $customQuestion->id, 'input' => $request->only('question', 'is_required')]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($c = self::guard($request, 'edit custom question', self::REDIRECT_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            if (!$this->isOwner($customQuestion)) {
                Log::warning("$sig not owner", ['user' => Auth::id()]);
                return defaultPermissionDenial($request, new AuthorizationException, $sig);
            }

            $t = microtime(true);
            $validator = Validator::make($request->all(), ['question' => 'required']);
            $this->logExecutionTime($t, "$sig::validate", 'completed');
            if ($validator->fails()) {
                $msg = $validator->errors()->first();
                Log::warning("$sig validation failed", ['error' => $msg]);
                return redirect()->back()->with('error', $msg);
            }

            try {
                $t = microtime(true);
                DB::transaction(function () use ($request, $customQuestion) {
                    foreach (['question', 'is_required'] as $field) {
                        $customQuestion->$field = $request->input($field);
                    }
                    $customQuestion->save();
                    Log::info('update: saved', ['id' => $customQuestion->id]);
                });
                $this->logExecutionTime($t, "$sig::transaction", 'completed');

                return redirect()->route(self::REDIRECT_ROUTE)->with('success', __('Question successfully updated.'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $sig);
            }
        });
    }

    public function destroy(CustomQuestion $customQuestion, Request $request): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $cls = static::class;
        $sig = "$cls::$action";
        return $this->measureProfile(function () use ($customQuestion, $request, $action, $sig) {
            Log::info("$sig start", ['id' => $customQuestion->id]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($c = self::guard($request, 'delete custom question', self::REDIRECT_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            if (!$this->isOwner($customQuestion)) {
                Log::warning("$sig not owner", ['user' => Auth::id()]);
                return defaultPermissionDenial($request, new AuthorizationException, $sig);
            }

            try {
                $t = microtime(true);
                DB::transaction(function () use ($customQuestion) {
                    $customQuestion->delete();
                    Log::info('destroy: deleted', ['id' => $customQuestion->id]);
                });
                $this->logExecutionTime($t, "$sig::transaction", 'completed');

                return redirect()->route(self::REDIRECT_ROUTE)->with('success', __('Question successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $sig);
            }
        });
    }

    private function isOwner(CustomQuestion $question): bool|RedirectResponse
    {
        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) {
            return $userOrRedirect;
        }
        $user = $userOrRedirect;
        return $question[DatabaseConstants::COL_TABLE_CREATOR] === $user?->creatorId();
    }
}
